### ISSUE TITLE
[Bug] Display corruption occurs when WebRender is enabled

### ISSUE THREAD

**Issue by jomegu** [mozilla-mobile/fenix#16469]
[comment]: # (Please do your best to search for duplicate issues before filing a new issue so we can keep our issue board clean)
[comment]: # (Every issue should have the exact bug and steps to reproduce described in it. Please do not file feedback list tickets as it is difficult to parse them and address their individual points)
[comment]: # (Read https://github.com/mozilla-mobile/fenix#i-want-to-file-an-issue for more information)

## Steps to reproduce

1. Go to `about:config`
2. Click on the search input
3. Write something on the search input (in my case - screenshot - I wrote `Webrender`)

### Expected behavior

The cursor in the input to flash and the text to appear without any corruption.

### Actual behavior

Display corruption occurs, showing weird rectangles when the cursor is flashing (after step 2):
![Screenshot_20201109-201331_Firefox Nightly](https://user-images.githubusercontent.com/6190345/98613693-b2aa5400-22ee-11eb-8404-acda535ddb59.jpg)
and garbled text (after step 3):
![Screenshot_20201109-201356_Firefox Nightly](https://user-images.githubusercontent.com/6190345/98613695-b3db8100-22ee-11eb-8ece-c8b02e3da8a3.jpg)

### Workaround

By setting `gfx.webrender.force-disabled` to `true`, the issue does not occur, as it falls back to using OpenGL instead of WebRender.

### Device information

* Android device: Samsung Galaxy S20 Ultra (Exynos 990 / Mali-G77 MP11)
* Fenix version: Firefox Nightly 84.0a1; Build ID20201106093443

### Extra notes

This issue occurs on almost every website, especially when writing on inputs, but also when interacting with other elements. In the screenshot below, the text became like that after I clicked the "Copy text to clipboard" button.
![Screenshot_20201109-211413_Firefox Nightly](https://user-images.githubusercontent.com/6190345/98613696-b3db8100-22ee-11eb-931e-bd95bfe22f24.jpg)

When using Firefox Nightly 84.0a1 Build ID20201105045247, the issue does not occur (WebRender is enabled according to the `about:support` page).

Not sure if it helps, but here is the text from the `about:support` page for the [20201105045247 build](https://github.com/mozilla-mobile/fenix/files/5514146/firefox20201105045247.txt) (the one that works as expected) and the [20201106093443 build](https://github.com/mozilla-mobile/fenix/files/5514147/firefox20201106093443.txt) (the one where the issue occurs).

Using an Xperia XZ2 Compact (Qualcomm SDM845 / Adreno 630) with the same Build ID20201106093443, this issue does not occur (WebRender is enabled according to the `about:support` page).

**Comment by kbrosnan** [comment]
Moved to bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1676474

Change performed by the [Move to Bugzilla add-on](https://addons.mozilla.org/en-US/firefox/addon/move-to-bugzilla/).

**Comment by jamienicol** [comment]
Hi @jomegu, thanks for the report! Would you please be able to enable `gfx.webrender.debug.gpu-cache` to true. This should show a grey rectangle with lots of red lines in the top of the screen. As you browse, red lines appear and dissappear. Can you see if you can reproduce the issue with this visible, and if possible take a screen recording of it?

If you can't take a screen recording, then try to describe what happens to me? At the moment the bug happens, do lots of red lines appear or dissappear? does the grey rectangle increase in size? Thanks!

**Comment by jamienicol** [comment]
Since you know this is a regression,  it would be *invaluable* if you could run mozregression to find out what caused it. There are some instructions for that here: https://mozilla.github.io/mozregression/quickstart.html

The command line in this case would be something like `mozregression --app gve --good 2020-01-01`. (A few days before the first version of Firefox with the bug). Then you test each version and answer good or bad, and eventually it should tell you what caused the bug. If you need any help with that please let me know!

**Comment by jomegu** [comment]
When I set `gfx.webrender.debug.gpu-cache` to `true` the issue does not occur. I recorded the [video](https://www.youtube.com/watch?v=IM-NbLGhwYY) **(be warned that this is probably dangerous for people prone to photosensitive epilepsy)** anyway, where I show it happening before setting that flag, then working as expected when the flag is turned on, and then the issue occurring again after I turn it back to `false`.

About mozregression, I couldn't get it to properly run, as I kept getting this kind of errors:
``` ERROR : Unable to start the application (error: args: adb wait-for-device shell run-as org.mozilla.geckoview_example cp -R /data/local/tmp/C:\Users\joaol\AppData\Local\Temp\tmpsjekqmpu /data/data/org.mozilla.geckoview_example/test_root/tmp1zjpqbo0.mozrunner, exitcode: 1, stdout: cp: bad '/data/local/tmp/C:UsersjoaolAppDataLocalTemptmpsjekqmpu': No such file or directory)```
The best that I could do was using the "Run a single build" function, and the build from 2020-11-05 (5278acfcd834) works correctly, while the one from 2020-11-06 (0e95e169ef40) doesn't.

**Comment by jamienicol** [comment]
Thanks! It's a shame that mozregression isn't cooperating so we can narrow it down even further, but even that is really helpful. It narrows it down to this range: https://hg.mozilla.org/mozilla-central/pushloghtml?fromchange=5278acfcd8344d6cd0a3f5bfb6cab6d68a9e93bc&tochange=81a3ef82469bf0bfd2e85086f3dfa3f91724f323

Out of those, `Glenn Watson — Bug 1675159 - Pt 2 - Expose dirty rects to start_compositing` seems the most suspicious to me. Could you please try setting `gfx.webrender.max-partial-present-rects` to zero in about:config. (it should currently be 1). **Then restart**, then see if the issue still reproduces?

**Comment by jamienicol** [comment]
If mozregression's `--launch` seems to work, could you try:

`mozregression --app gve --launch 93e746226b8feb5a5e6a643037ef34496caa26bf` (should be bad)

and

`mozregression --app gve --launch 20478a1a36cbf1681fca55a549833f810368984b` (should be good)

to confirm whether that is indeed to the commit which caused the bug.

**Comment by jomegu** [comment]
Both of those install the same build (`81a3ef82469b-shippable--mozilla-central--arm--geckoview_example.apk`), where the issue occurs. Even using `dff736842d4e86cf0532c01d10b3438a7fb6e80b` which seems to be the oldest from that list also installs that build.

The good news are that yesterday, after posting my comment, I noticed that Firefox Nightly got an update from Play Store and this issue is no longer occurring. I already got another update today and it's still working as expected.
I guess it's probably not worth the effort of finding the root case of this issue, but the change that fixed this issue should have been between the build `0e95e169ef40` (bad) and the build `e40cc6272439` (good).

**Comment by jamienicol** [comment]
That's good to hear the problem has gone away, although it's slightly concerning that we don't know why. I can't see anything in that range which is really obvious.

I don't want to take up too much more of your time, but it would be really helpful to find out a wee bit more information, otherwise we risk this bug coming back again without us knowing.

> Both of those install the same build (81a3ef82469b-shippable--mozilla-central--arm--geckoview_example.apk), where the issue occurs. Even using dff736842d4e86cf0532c01d10b3438a7fb6e80b which seems to be the oldest from that list also installs that build.

Are you using the mozregression GUI? I think if you select the "repository" as "autoland" that might fix it. (or on the command line add `--repo autoland`)

If you are then able to run a specific build, if you could confirm which build introduced the bug and possibly even narrow down which one fixed it that would be so helpful. Thanks you so much for all your help with this.

**Comment by jomegu** [comment]
Setting the repo to `autoland` gives me this error:
```
0:01.40 INFO: 93e746226b8feb5a5e6a643037ef34496caa26bf is not a release, assuming it's a hash...
0:05.34 ERROR: Unable to find build info using the taskcluster route 'gecko.v2.autoland.revision.93e746226b8feb5a5e6a643037ef34496caa26bf.mobile.android-api-16-opt'
```
(both on the command line and in the GUI

**Comment by jamienicol** [comment]
Ugh, that's really annoying that it isn't cooperating. Thanks for all your help anyway. Not much we can really do here for now, but please let me know if you do see this issue again!

**Comment by jomegu** [comment]
Unfortunately I recently (Friday, I think) started having this issue again on Firefox Nightly (same behaviour, only when using WebRender, and goes away when enabling `gfx.webrender.debug.gpu-cache`).
Today I had some time to use mozregression (with the autoland repo) and it seems the change that caused the issue is [between d22dc65be538930b205704bb8026dd7dfd51a833 (good) and 98c926d9289f5c45565a3a157971b46f98210133 (bad)](https://hg.mozilla.org/integration/autoland/pushloghtml?fromchange=d22dc65be538930b205704bb8026dd7dfd51a833&tochange=98c926d9289f5c45565a3a157971b46f98210133).
Unfortunately I couldn't get any build between those two changesets, so I can't know for sure which one of it is causing the issue.

**Comment by jamienicol** [comment]
Hmm, this is unfortunate.

Could you try reseting gfx.webrender.debug.gpu-cache and all other prefs, so that it *does* reproduce. Then try setting `gfx.webrender.max-partial-present-rects` to 0 (it should be 1 currently). Restart, and then see if it reproduces the bug.

Again, reset that pref, then try setting `gfx.webrender.debug.texture-cache` to true, restart, and see if that works?

Could you also please try the firefox beta version (with all webrender prefs set to their default), and see if that is buggy?

**Comment by jomegu** [comment]
By starting in a state where the issue occurs, then setting `gfx.webrender.max-partial-present-rects` to 0 and restarting the app, the issue no longer occurs.
After setting it back to 1 and restarting the app, so that the issue occurs again, when I set `gfx.webrender.debug.texture-cache` to true and restart the app, the issue no longer occurs.
On a fresh install of Firefox Beta (84.0b3, build 20201119195818) the issue does not occur. (WebRender is being used, according to about:support)

**Comment by jamienicol** [comment]
Thanks. Okay, that confirms it's partial present related, as we suspected based on the fact that the GPU cache debug view precents the bug. On nightly could you try reseting `gfx.webrender.max-partial-present-rects` back to the default of 1 (and everything else to default too), and set `gfx.webrender.allow-partial-present-buffer-age` to `false`. Then restart again and see if it reproduces? The partial present feature has 2 parts to it, so this will tell us which part is causing the issue.

Are you able to capture a logcat while you are reproducing the bug? If so, it would be much appreciated if you could attach that here.

Thanks!

**Comment by jamienicol** [comment]
@jomegu  Unfortunately there's been delays updating the beta version, so the version you tested is too old to be affected by the bug. A new version is being rolled out today and tomorrow, once you get that update could you please check again? That would be incredibly helpful, as we'll need to decide whether to revert this change from the beta version at the last minute.

**Comment by jomegu** [comment]
On Nightly, I can confirm that after resetting `gfx.webrender.max-partial-present-rects` and setting `gfx.webrender.allow-partial-present-buffer-age` to false, the issue also does not occur.

I installed the beta 4 from the [releases page](https://github.com/mozilla-mobile/fenix/releases/tag/v84.0.0-beta.4) and I can confirm that the issue also occurs there.

I captured a logcat while reproducing the issue on Firefox Beta (since it's a fresh install):
[log.txt](https://github.com/mozilla-mobile/fenix/files/5661133/log.txt)
